{% extends 'frontend/base.html' %}

{% block title %}Variant{% endblock %}

{% block css %}
  <!-- Allows text wrapping in Tabulator title text. Reference: https://github.com/olifolkerd/tabulator/issues/988 -->
  <style>
    .tabulator .tabulator-header .tabulator-col .tabulator-col-content .tabulator-col-title {
      white-space: normal;
      text-overflow: clip;
    }
    .tooltip a {
      text-decoration: underline;
    }
  </style>
  <link rel="stylesheet" type="text/css"
        href="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.css"/>
  <link rel="stylesheet"
        href="https://cdn.jsdelivr.net/npm/tabulator-tables@4.4.3/dist/css/bootstrap/tabulator_bootstrap4.min.css">
{% endblock %}

{% block content %}
  <div class="container-fluid">

    <div class="row padtop">
      <div class="col-sm-10">
        <div id="home-searchbox">
          <form id="variant-search">
            <div class="input-group">
              <div class="input-group-prepend">
                  <a href="/" class="btn btn-secondary btn-sm" role="button" aria-pressed="true"><span class="fa fa-home"></span></a>
              </div>
              <input id="query-field" name="query-field" autocomplete="off" class="form-control"
                      type="text"
                      placeholder="Search for a variant, e.g. chr19:488506 or rs10424907" autofocus/>
              <div class="input-group-append">
                <button id="navigate-variant" class="btn btn-secondary" type="submit"><span
                    class="fa fa-search"></span></button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>

  <div class="container-fluid">
    <div class="row padtop">
      <div class="col">
        <h1>Variant: chr{{ chrom }}:{{ "{:,}".format(pos) }} {{ ref }}/{{ alt }} ({{ rsid }})</h1>
      </div>
    </div>
  </div>

  <div class="container-fluid">
    <div class="row justify-content-start">

      <div class="col-12 col-lg-4" style="margin-bottom:0.8em">
        <form class="grouping" id="grouping-buttons">
          <div class="btn-group btn-group-toggle" data-toggle="buttons">
            <span class="d-inline-block" tabindex="0" data-toggle="tooltip"
                  title="Reorder the x-axis by one of these listed categories" data-placement="top">
              <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-exchange-alt"></span> Group by: </button>
            </span>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Group eQTLs by tissues, sorted alphabetically">
              <input type="radio" name="group-options" autocomplete=off value="tissue"> Tissue
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Group eQTLs by systems as defined by the GTEx project, sorted alphabetically">
              <input type="radio" name="group-options" autocomplete=off value="system"> System
            </label>
            <label class="btn btn-secondary active" data-toggle="tooltip" data-placement="top"
                   title="Group eQTLs by gene, sorted by the position of the genes' transcription start sites">
              <input type="radio" name="group-options" autocomplete=off value="symbol"> Gene
            </label>
          </div>
        </form>
      </div>

      <div class="col-12 col-lg-4" style="margin-bottom:0.8em">
        <form class="yaxis-display" id="transform-y">
          <div class="btn-group btn-group-toggle" data-toggle="buttons">
            <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true"
                  title="Switches the variable plotted on the Y-axis between -log<sub>10</sub>(P-value) and Normalized Effect Size (NES). Triangles indicate eQTLs for upregulation (pointing up) or downregulation (pointing down) of gene expression with P-values < 0.05.">
              <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-arrows-alt-v"></span> Show on Y-Axis: </button>
            </span>
            <label class="btn btn-secondary active" data-toggle="tooltip" data-placement="top" data-html="true"
                   title="Display -log<sub>10</sub>(P-values) on the Y-axis">
              <input type="radio" name="y-options" autocomplete=off value="log_pvalue"> P-value
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" data-html="true"
                   title="Displays Normalized Effect Size (NES) on the Y-axis. See <a href='https://www.gtexportal.org/home/documentationPage' target='_blank'>the GTEx Portal</a> for an explanation of NES.">
              <input type="radio" name="y-options" autocomplete=off value="beta"> Effect Size
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" data-html="true"
                   title="Displays <a href='https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006646' target='_blank'>DAP-G</a> Posterior Inclusion Probabilities (PIP) on the Y-axis.<br>Cluster 1 denotes the cluster of variants (in LD with each other) with the strongest signal; cluster 2 denotes the set of variants with the next strongest signal; and so on.">
              <input type="radio" name="y-options" autocomplete=off value="pip"> PIP
            </label>
          </div>
        </form>
      </div>

      <div class="col-12 col-lg-4" style="margin-bottom:0.8em">
        <form class="label-display" id="toggle-labels">
          <div class="btn-group btn-group-toggle" data-toggle="buttons">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" data-html="true"
                title="Toggles labels for the most significant eQTLs. Significance is determined by P-values. <b>Will only label variants more significant than 10<sup>-20</sup></b>. When viewing <b>Effect Size (NES)</b>, show either 5 or 50 eQTLs with the largest absolute effects <b>only</b> if they are also more significant than 10<sup>-20</sup></b>.">
            <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-tag"></span> Label: </button>
          </span>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" title="Turn off all labels">
              <input type="radio" name="label-options" autocomplete=off value="0"> None
            </label>
            <label class="btn btn-secondary active" data-toggle="tooltip" data-placement="top" data-html="true"
                   title="If viewing P-values, Add labels to the 5 eQTLs with the most significant P-values <b>if they are more significant than 10<sup>-20</sup></b>. If viewing Effect Sizes, choose the eQTLs with the 5 largest absolute effect sizes and only label those with P-value more significant than 10<sup>-20</sup>.">
              <input type="radio" name="label-options" autocomplete=off value="5"> Top 5
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" data-html="true"
                   title="If viewing P-values, add labels to the 50 eQTLs with the most significant P-values <b>if they are most significant than 10<sup>-20</sup></b>. If viewing Effect Sizes, choose the eQTLs with the 50 largest absolute effect sizes and only label those with P-value more significant than 10<sup>-20</sup>.">
              <input type="radio" name="label-options" autocomplete=off value="50"> Top 50
            </label>
          </div>
        </form>
      </div>
    </div>

    <!-- Filters all points by TSS. Applies a region filter which is symmetric around the current point -->
    <div class="row justify-content-start">
      <div class="col-12" style="margin-bottom:0.8em">
        <form class="tss-range" id="tss-both-range">
          <div class="btn-group btn-group-toggle" data-toggle="buttons">
            <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true"
                  title="Display eQTLs for genes <b>only</b> if their Transcription Start Sites (TSS's) are within the selected distance from this variant.">
              <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;">
                <span class="fa fa-arrows-alt-h"></span> Filter by TSS Distance
              </button>
            </span>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 20000, 1] | max) }}-{{ '{:,}'.format(pos + 20000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="20000"> ±20kb
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 50000, 1] | max) }}-{{ '{:,}'.format(pos + 50000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="50000"> ±50kb
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 100000, 1] | max) }}-{{ '{:,}'.format(pos + 100000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="100000"> ±100kb
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 200000, 1] | max) }}-{{ '{:,}'.format(pos + 200000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="200000"> ±200kb
            </label>
            <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 500000, 1] | max) }}-{{ '{:,}'.format(pos + 500000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="500000"> ±500kb
            </label>
            <label class="btn btn-secondary active" data-toggle="tooltip" data-placement="top"
                   title="Show eQTLs for genes with TSS in the range chr{{ chrom }}:{{ '{:,}'.format([pos - 1000000, 1] | max) }}-{{ '{:,}'.format(pos + 1000000) }}">
              <input type="radio" name="tss-options" autocomplete=off value="1000000"> ±1mb
            </label>
          </div>
        </form>
      </div>
    </div>
  </div>

  <div class="container-fluid">
    <div id="lz-plot">
    </div>
  </div>

  <div class="container-fluid">
    <div class="row justify-content-center">
      <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top"
        title="The red dotted line indicates the current variant's position with respect to the scale on the gene track.">
        <span class="badge badge-pill badge-secondary" style="pointer-events: none;"><span class="fa fa-question"></span> Red dotted line </span>
      </span>
    </div>
  </div>


  <div class="container-fluid">
    <div class="row">
      <div class="col">
        <button class="btn btn-secondary" type="button" data-toggle="collapse" data-target="#variantInformation"
                aria-expanded="true" aria-controls="variantInformation">
          Variant Information from Sequence Genotype <span class="fa fa-caret-down"></span>
        </button>
        <div class="collapse show" id="variantInformation">
          <div class="card-group">
            <div class="card variant-information-grouped-card">
              <div class="card-body">
                <dl class="variant-information-dl">
                  {% if ac is not none and an is not none %}
                    <dt>Allele count / total</dt>
                    <dd>{{ ac }} / {{ an }}</dd>
                  {% else %}
                    <dt>Note:</dt>
                    <dd>Allele information not found</dd>
                  {% endif %}

                  {% if af is not none %}
                    <dt>Allele frequency</dt>
                    <dd>{{ af }}</dd>
                    </dl>
                  {% endif %}
              </div>
            </div>

            <div class="card variant-information-grouped-card">
              <div class="card-body">
                <dl class="variant-info-middle">
                  {% if top_gene is not none %}
                    <dt>Top gene</dt>
                    <dd><i>{{ top_gene }}</i></dd>
                  {% endif %}
                  {% if top_tissue is not none %}
                    <dt>Top tissue</dt>
                    <dd>{{ top_tissue }}</dd>
                  {% endif %}
                </dl>
              </div>
            </div>

            <div class="card variant-information-grouped-card">
              <div class="card-body">
                <dl class="variant-information-dl">
                  {% if rsid is not none %}
                    <dt>
                      rsid
                    </dt>
                    <dd>
                      {{ rsid }}
                    </dd>
                  {% endif %}
                  <dt>
                    {{ "Overlapping" if is_inside_gene else "Nearest" }} gene(s)
                  </dt>
                  <dd>
                    <i>
                      {% for gene in nearest_genes %}
                        <span class="text-with-definition" title="{{ gene.ensg }}">{{ gene.symbol }}</span>
                        {%- if not loop.last -%},{% endif %}
                      {% else %}
                        (no genes found)
                      {% endfor %}
                    </i>
                  </dd>
                </dl>
              </div>
            </div>
          </div>

          <div class="card">
            <div class="card-body">
              <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top"
                    title="External links for more information about this variant">
                <button class="btn btn-sm btn-secondary" style="pointer-events: none;"><span
                    class="fa fa-secondary-circle"></span><span class="fa fa-info-circle"></span> Variant info </button>
              </span>
              {% if ref is not none and alt is not none %}
                <a href="https://bravo.sph.umich.edu/freeze5/hg38/variant/{{ chrom }}-{{ pos }}-{{ ref }}-{{ alt }}" target="_blank"
                    class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
                    data-placement="top" data-html=true
                    title="Variant data from NHLBI's TOPMed program, containing 463 million variants observed in 62,784 individuals in data freeze 5. <b>Requires Google login</b>">
                  BRAVO <span class="fa fa-external-link-alt"></span> </a>
                <a href="https://gtexportal.org/home/snp/chr{{ chrom }}_{{ pos }}_{{ ref }}_{{ alt }}_b38" target="_blank"
                    class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
                    data-placement="top"
                    title="Variant data from the Genotype-Tissue Expression project, containing expression data, histology images, and in-depth expression data analysis">
                  GTEx Portal <span class="fa fa-external-link-alt"></span> </a>
                <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&highlight=hg38.chr{{ chrom }}%3A{{ pos }}-{{ pos }}&position=chr{{ chrom }}%3A{{ pos - 25 }}-{{ pos + 25 }}" target="_blank"
                    class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
                    data-placement="top" title="The UC Santa Cruz Genome Browser"> UCSC <span
                    class="fa fa-external-link-alt"></span></a>
                <a href="https://gnomad.broadinstitute.org/variant/chr{{ chrom }}-{{ pos }}-{{ ref }}-{{ alt }}?dataset=gnomad_r3" target="_blank"
                    class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
                    data-placement="top"
                    title="The Genome Aggregation Database (v3) at the Broad Institute, containing variant data from 71,702 sequenced genomes">
                  gnomAD <span class="fa fa-external-link-alt"></span></a>
              {% endif %}
              {% if rsid is not none %}
                <a href="https://www.ncbi.nlm.nih.gov/snp/{{ rsid }}" target="_blank" class="btn btn-secondary btn-sm" role="button"
                    aria-pressed="true" data-toggle="tooltip" data-placement="top"
                    title="Reference SNP Report from the National Center for Biotechnology Information (NCBI)"> dbSNP
                  <span class="fa fa-external-link-alt"></span> </a>
                <a href="http://pheweb.sph.umich.edu/go?query={{ rsid }}" target="_blank" class="btn btn-secondary btn-sm" role="button"
                    aria-pressed="true" data-toggle="tooltip" data-placement="top"
                    title="PheWeb summary of association results from 1,448 electronic health record-derived phenotypes tested against up to ~6,000 cases and ~18,000 controls with genotyped and imputed samples from the Michigan Genomics Initiative">
                  MGI <span class="fa fa-external-link-alt"></span></a>
                <a href="http://pheweb.sph.umich.edu/SAIGE-UKB/go?query={{ rsid }}" target="_blank" class="btn btn-secondary btn-sm"
                    role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top"
                    title="PheWeb summary of association results from the UK Biobank, with up to ~78k cases and ~409k controls, with binary outcomes analyzed with the SAIGE software">
                  UKB-SAIGE <span class="fa fa-external-link-alt"></span></a>
                <a href="http://big.stats.ox.ac.uk/go?query={{ rsid }}" target="_blank" class="btn btn-secondary btn-sm" role="button"
                    aria-pressed="true" data-toggle="tooltip" data-placement="top"
                    title="Summary of 3,144 GWAS of Brain Imaging Derived Phenotypes (IDPs) in 9,707 participants from the UK Biobank, analyzed with the BGENIE software">
                  UKB-Oxford BIG <span class="fa fa-external-link-alt"></span></a>
              {% endif %}
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>

  <div class="container-fluid">
    <div id="table" style="width:90%;margin:0 auto">
    </div>
  </div>

  <!-- Supporting information about this project -->
  <div class="container-fluid">
    <div class="card">
      <div class="card-body">
        <a href="https://www.gtexportal.org/home/datasets" class="badge badge-pill badge-secondary"
           data-toggle="tooltip" data-placement="top"
           title="eQTL data was obtained from the GTEx Project, v8, with 49 tissues in up to 670 subjects with both genotypes and expression data">
          <span class="fa fa-file-text-o"></span> Data source <span class="fa fa-external-link-alt"></span> </a>
        <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" data-html="true"
              title="Normalized Effect Sizes (NES) are defined as the effect of one allele on the inverse-normalized expression level of the associated gene. Go <a href='https://www.gtexportal.org/home/documentationPage' target='_blank'>here</a> for more details from the GTEx Portal. <br>
                     Posterior Inclusion Probabilities (PIP) were calculated using DAP-G (Wen et al. 2017). See <a href='https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006646' target='_blank'>the associated paper in PLoS Genetics</a> for details.">
        <span class="badge badge-pill badge-secondary" style="pointer-events: none;">
          <span class="fa fa-info-circle"></span> Definitions </span> </span>
        <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" data-html="true"
              title="Created by Alan Kwong, Mukai Wang, Andy Boughton, Peter VandeHaar, and Hyun Min Kang. Source code can be found on <a href=https://github.com/statgen/pheget/>GitHub</a>.">
        <span class="badge badge-pill badge-secondary" style="pointer-events: none;"><span
            class="fa fa-lightbulb"></span> Credits</span>
      </span>
      </div>
    </div>
  </div>
{% endblock %}


{% block javascript %}
  {# These files must be included to make a LocusZoom plot #}
  <script src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.vendor.min.js"
          type="application/javascript"></script>
  <script src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.app.min.js"
          type="application/javascript"></script>
  <script src="https://cdn.jsdelivr.net/npm/tabulator-tables@4.4.3/dist/js/tabulator.min.js"
          type="text/javascript"></script>

  <script type="application/javascript"
          src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/ext/lz-dynamic-urls.min.js"></script>

  <script src="{{ url_for('static', filename='js/common.js') }}" type="application/javascript"></script>
  <script src="{{ url_for('static', filename='js/variant.js') }}" type="application/javascript"></script>
  <script src="{{ url_for('static', filename='js/searchbox.js') }}" type="application/javascript"></script>
  <script type="application/javascript">
      const table = makeTable('#table');
      const [plot, datasources] = makePhewasPlot({{chrom|tojson}}, {{pos|tojson}}, '#lz-plot');
      plot.subscribeToData(['phewas:log_pvalue', 'phewas:gene_id', 'phewas:tissue', 'phewas:system', 'phewas:symbol', 'phewas:beta', 'phewas:stderr_beta', 'phewas:pip'], updateTable.bind(null, table));

      // Use omnisearch (https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch) to parse rsids
      document.getElementById('variant-search').addEventListener('submit', function () {
          event.preventDefault();
          const variant = document.getElementById('query-field').value;
          if (!variant) {
              return;
          }
          parseSearch(variant);
      });

      // On hitting toggle buttons, run JavaScript to update plot views
      document.forms['grouping-buttons'].elements['group-options'].forEach(function (form) {
          form.onchange = function () {
              groupByThing(plot, this.value);
          };
      });

      document.forms['transform-y'].elements['y-options'].forEach(function (form) {
          form.onchange = function () {
              switchY(plot, table, this.value);
          };
      });

      document.forms['toggle-labels'].elements['label-options'].forEach(function (form) {
          form.onchange = function () {
              labelTopVariants(plot, this.value);
          };
      });

      // Filters displayed points only if they are within +/- the given distance to the current variant
      // The position of this variant does not change, and is stored at plot.state.position
      document.forms['tss-both-range'].elements['tss-options'].forEach(function (form) {
          form.onchange = function () {
              buttonValue = +this.value;
              // Change overall start and end state to update displayed gene track range
              // And update the stored tss distance fields
              plot.applyState({
                  maximum_tss_distance: buttonValue,
                  minimum_tss_distance: -buttonValue,
                  start: Math.max(plot.state.position - buttonValue, 1),
                  end: plot.state.position + buttonValue
              });
          };
      });

  </script>
{% endblock %}
