<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/locuszoom@0.10.0-beta.1/dist/locuszoom.css"
        type="text/css"/>

  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
        integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"
        integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">
  <link rel="stylesheet"
        href="https://cdn.jsdelivr.net/npm/tabulator-tables@4.4.3/dist/css/bootstrap/tabulator_bootstrap4.min.css">
  <title>PheGET - Single Variant View</title>

  <style>
  .padtop {
    padding-top: 20px;
  }

  .text-with-definition {
    text-decoration: dotted underline;
  }

  /* Formats dt/dd to display on the same line */
  dl.inline-flex {
    display: flex;
    flex-flow: row;
    flex-wrap: wrap;
    width: 360px;
    overflow: visible;
  }
  dl.inline-flex dt {
    flex: 0 0 50%;
    text-overflow: ellipsis;
    overflow: hidden;
  }
  dl.line-flex dd {
    margin-left: auto;
    text-align: left;
    text-overflow: ellipsis;
    overflow: hidden;
    flex: 0 0 50%
  }
  </style>

</head>
<body>

<div class="container-fluid">
  <div class="row padtop">
    <div class="col-sm-10">
      <div id="home-searchbox">
        <form id="variant-search">
          <div class="input-group">
            <input id="query-field" name="query-field" autocomplete="off" class="form-control"
                   type="text"
                   placeholder="Search for a variant, e.g. chr19:488506 or rs10424907" autofocus/>
            <div class="input-group-append">
              <button id="navigate-variant" class="btn btn-primary" type="submit">Search <span class="fa fa-search"></span></button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  <center>
  <a href="/" class="btn btn-primary btn-sm" role="button" aria-pressed="true"><span class="fa fa-home"></span> Back to PheGET Homepage</a>
  </center>
  <br>

  <h1>Variant: chr{{ chrom }}:{{ "{:,}".format(pos) }} {{ ref }}/{{ alt }} ({{ rsid }})</h1>

  <div class="row">
    <div class="col">
      <button class="btn btn-info" type="button" data-toggle="collapse" data-target="#variantInformation" aria-expanded="true" aria-controls="variantInformation">
        Variant Information from Sequence Genotype <span class="fa fa-caret-down"></span>
      </button>
      <div class="collapse show" id="variantInformation">

        <div class="card-group">

          <div class="card">
            <div class="card-body">
              <dl class="inline-flex">
                {% if ac is not none and an is not none %}
                    <dt>Allele count / total</dt>
                    <dd>{{ ac }} / {{ an }}</dd>
                {% else %}
                    <dt>Note:</dt>
                    <dd>Allele information not found</dd>
                {% endif %}

                {% if af is not none %}
                <dt>Allele frequency</dt>
                  <dd>{{ af }}</dd>
                </dl>
                {% endif %}
            </div>
          </div>

          <div class="card">
            <div class="card-body">
              <dl class="inline-flex">
                {% if top_gene is not none %}
                <dt>Top gene</dt>
                  <dd><i>{{ top_gene }}</i></dd>
                {% endif %}
                {% if top_tissue is not none %}
                <dt>Top tissue</dt>
                  <dd>{{ top_tissue }}</dd>
                {% endif %}
              </dl>
            </div>
          </div>

          <div class="card">
            <div class="card-body">
              <dl class="inline-flex">
                {% if rsid is not none %}
                <dt>
                 rsid
                </dt>
                <dd>
                 {{ rsid }}
                </dd>
                {% endif %}
                <dt>
                 {{ "Overlapping" if is_inside_gene else "Nearest" }} gene(s):
                </dt>
                <dd>
                  <i>
                 {% for gene in nearest_genes %}
                 <span class="text-with-definition" title="{{ gene.ensg }}">{{ gene.symbol }}</span>
                 {%- if not loop.last -%},{% endif %}
                 {% else %}
                  (no genes found)
                 {% endfor %}
                 </i>
                </dd>
              </dl>
            </div>
          </div>
        </div>

            <div class="card">
              <div class="card-body">
                <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" title="External links for more information about this variant">
                  <button class="btn btn-sm btn-secondary" style="pointer-events: none;"><span class="fa fa-info-circle"></span> Variant info </button>
                </span>
                {% if ref is not none and alt is not none %}
                <a href="https://bravo.sph.umich.edu/freeze5/hg38/variant/{{ chrom }}-{{ pos }}-{{ ref }}-{{ alt }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" data-html=true title="Variant data from NHLBI's TOPMed program, containing 463 million variants observed in 62,784 individuals in data freeze 5. <b>Requires Google login</b>"> BRAVO <span class="fa fa-external-link"></span> </a>
                <a href="https://gtexportal.org/home/snp/chr{{ chrom }}_{{ pos }}_{{ ref }}_{{ alt }}_b38" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="Variant data from the Genotype-Tissue Expression project, containing expression data, histology images, and in-depth expression data analysis"> GTEx Portal <span class="fa fa-external-link"></span> </a>
                <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&highlight=hg38.chr{{ chrom }}%3A{{ pos }}-{{ pos }}&position=chr{{ chrom }}%3A{{ pos - 25 }}-{{ pos + 25 }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="The UC Santa Cruz Genome Browser"> UCSC <span class="fa fa-external-link"></span></a>
                <a href="https://gnomad.broadinstitute.org/variant/{{ chrom }}-{{ pos }}-{{ ref }}-{{ alt }}?dataset=gnomad_r3" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="The Genome Aggregation Database (v3) at the Broad Institute, containing variant data from 71,702 sequenced genomes"> gnomAD <span class="fa fa-external-link"></span></a>
                {% endif %}
                {% if rsid is not none %}
                <a href="https://www.ncbi.nlm.nih.gov/snp/{{ rsid }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="Reference SNP Report from the National Center for Biotechnology Information (NCBI)"> dbSNP <span class="fa fa-external-link"></span> </a>
                <a href="http://pheweb.sph.umich.edu/go?query={{ rsid }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="PheWeb summary of association results from 1,448 electronic health record-derived phenotypes tested against up to ~6,000 cases and ~18,000 controls with genotyped and imputed samples from the Michigan Genomics Initiative"> MGI <span class="fa fa-external-link"></span></a>
                <a href="http://pheweb.sph.umich.edu/SAIGE-UKB/go?query={{ rsid }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="PheWeb summary of association results from the UK Biobank, with up to ~78k cases and ~409k controls, with binary outcomes analyzed with the SAIGE software"> UKB-SAIGE <span class="fa fa-external-link"></span></a>
                <a href="http://big.stats.ox.ac.uk/go?query={{ rsid }}" class="btn btn-primary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top" title="Summary of 3,144 GWAS of Brain Imaging Derived Phenotypes (IDPs) in 9,707 participants from the UK Biobank, analyzed with the BGENIE software"> UKB-Oxford BIG <span class="fa fa-external-link"></span></a>
                {% endif %}
              </div>
            </div>

      </div>
    </div>
  </div>
</div>

<br>

<div class="container-fluid">
  <div class="row justify-content-start">

    <div class="col">
      <form action="#" method="post" class="grouping" id="grouping-buttons">
        <div class="btn-group btn-group-toggle" data-toggle="buttons">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" title="Reorder the x-axis by one of these listed categories" data-placement="top">
            <button type="button" class="btn btn-secondary" style="pointer-events: none;"> Group by: </button>
          </span>
          <label class="btn btn-primary" data-toggle="tooltip" data-placement="top" title="Group eQTLs by tissues, sorted alphabetically">
              <input type="radio" name="group-options" autocomplete=off id="group-tissue" value="tissue"> Tissue
          </label>
          <label class="btn btn-primary"  data-toggle="tooltip" data-placement="top" title="Group eQTLs by systems as defined by the GTEx project, sorted alphabetically">
              <input type="radio" name="group-options" autocomplete=off id="group-system" value="system"> System
          </label>
          <label class="btn btn-primary active" data-toggle="tooltip" data-placement="top" title="Group eQTLs by gene, sorted by the position of the genes' translation start sites">
              <input type="radio" name="group-options" autocomplete=off id="group-symbol" value="symbol"> Gene
          </label>
        </div>
      </form>
    </div>

    <div class="col-sm">
      <form action="#" method="post" class="yaxis-display" id="transform-y">
        <div class="btn-group btn-group-toggle" data-toggle="buttons">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true" title="Switches the y variable between -log<sub>10</sub>(P-value) and Normalized Effect Size (NES). Triangles indicate eQTLs for upregulation (pointing up) or downregulation (pointing down) of gene expression with P-values < 0.05.">
            <button type="button" class="btn btn-secondary" style="pointer-events: none;"> Show: </button>
          </span>
          <label class="btn btn-success active" data-toggle="tooltip" data-placement="top" data-html="true" title="Display -log<sub>10</sub>(P-values) on the Y-axis">
            <input type="radio" name="y-options" autocomplete=off id="show-log-pvalue" value="log_pvalue"> P-values
          </label>
          <label class="btn btn-success" data-toggle="tooltip" data-placement="top" data-html="true" title="Displays Normalized Effect Size (NES) on the Y-axis. See <a href='https://www.gtexportal.org/home/documentationPage'>the GTEx Portal</a> for an explanation of NES.">
            <input type="radio" name="y-options" autocomplete=off id="show-beta" value="beta"> Effect Size (NES)
          </label>
        </div>
      </form>
    </div>

    <div class="col-sm">
      <form action="#" method="post" class="label-display" id="toggle-labels">
        <div class="btn-group btn-group-toggle" data-toggle="buttons">
          <span class="d-inline=block" tabindex="0" data-toggle="tooltip" data-placement="top" data-html="true" title="Toggles labels for the most significant eQTLs. Significance is determined by P-values. <b>Will only label variants more significant than 10<sup>-20</sup></b>.">
           <button type="button" class="btn btn-secondary" style="pointer-events: none;"> Label: </button>
          </span>
          <label class="btn btn-info" data-toggle="tooltip" data-placement="top" title="Turn off all labels">
            <input type="radio" name="label-options" autocomplete=off id="label-none" value="0"> None
          </label>
          <label class="btn btn-info active" data-toggle="tooltip" data-placement="top" data-html="true" title="Add labels to the 5 eQTLs with the most significant P-values <b>if they are more significant than 10<sup>-20</sup></b>">
            <input type="radio" name="label-options" autocomplete=off id="label-5" value="5"> Top 5
          </label>
          <label class="btn btn-info"  data-toggle="tooltip" data-placement="top" data-html="true" title="Add labels to the 50 eQTLs with the most significant P-values <b>if they are most significant than 10<sup>-20</sup></b>">
            <input type="radio" name="label-options" autocomplete=off id="label-50" value="50"> Top 50
          </label>
        </div>
      </form>
    </div>

  </div>
</div>

<div id="plot"></div>

<div class="container-fluid">
  <div class="row justify-content-center">
    <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" title="The red dotted line indicates the current variant's position with respect to the scale on the gene track.">
      <span class="badge badge-pill badge-secondary" style="pointer-events: none;"><span class="fa fa-question"></span> Red dotted line </span>
    </span>
  </div>
</div>

<div id="table" style="width:90%;margin:0 auto"></div>

{# These files must be included to make a LocusZoom plot #}
<script src="https://cdn.jsdelivr.net/npm/locuszoom@0.10.0-beta.1/dist/locuszoom.vendor.min.js"
        type="application/javascript"></script>
<script src="https://cdn.jsdelivr.net/npm/locuszoom@0.10.0-beta.1/dist/locuszoom.app.min.js"
        type="application/javascript"></script>
<script src="https://cdn.jsdelivr.net/npm/tabulator-tables@4.4.3/dist/js/tabulator.min.js"
        type="text/javascript"></script>

<!-- Allows text wrapping in Tabulator title text. Reference: https://github.com/olifolkerd/tabulator/issues/988 -->
<style>
  .tabulator .tabulator-header .tabulator-col .tabulator-col-content .tabulator-col-title{
	white-space: normal;
	text-overflow: clip;
}
</style>

<script src="{{ url_for('static', filename='js/phewas_scatter.js') }}" type="application/javascript"></script>
<script src="{{ url_for('static', filename='js/searchbox.js') }}" type="application/javascript"></script>
<script type="application/javascript">
    const table = makeTable('#table');
    const [plot, datasources] = makePhewasPlot({{chrom|tojson}}, {{pos|tojson}}, '#plot');
    plot.subscribeToData(['phewas:log_pvalue', 'phewas:gene_id', 'phewas:tissue', 'phewas:system', 'phewas:symbol', 'phewas:beta', 'phewas:stderr_beta'], updateTable.bind(null, table));

    // Use omnisearch (https://portaldev.sph.umich.edu/api_internal_dev/v1/annotation/omnisearch) to parse rsids
    document.getElementById('variant-search').addEventListener('submit', function () {
        event.preventDefault();
        const variant = document.getElementById('query-field').value;
        if (!variant) {
            return;
        }
        parseSearch(variant);
    });

    // On hitting toggle buttons, run JavaScript to update plot views
    var groupVar = document.forms['grouping-buttons'].elements['group-options'];
    for (var i=0, len=groupVar.length; i<len; i++) {
      groupVar[i].onchange = function() {
        groupByThing(plot, this.value);
      }
    }

    var yaxisVar = document.forms['transform-y'].elements['y-options'];
    for (var j=0, len=yaxisVar.length; j<len; j++) {
      yaxisVar[j].onchange = function() {
        switchY(plot, table, this.value);
      }
    }

    var labelVar = document.forms['toggle-labels'].elements['label-options'];
    for (var k=0, len=labelVar.length; k<len; k++) {
      labelVar[k].onchange = function() {
        labelTopVariants(plot, this.value);
      }
    }

</script>

<!-- Minimum code to include Bootstrap 4 functionality -->
<script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>

<script>
// Activates Bootstrap tooltips
$(function () {
  $('[data-toggle="tooltip"]').tooltip()
})
</script>

<!-- Supporting information about this project -->
    <div class="card">
        <div class="card-body">
          <a href="https://www.gtexportal.org/home/datasets" class="badge badge-pill badge-secondary" data-toggle="tooltip" data-placement="top" title="eQTL data was obtained from the GTEx Project, v8, with 49 tissues in up to 670 subjects with both genotypes and expression data"> <span class="fa fa-file-text-o"></span> Data source <span class="fa fa-external-link"></span> </a>
          <a href="https://www.gtexportal.org/home/documentationPage" class="badge badge-pill badge-secondary" data-toggle="tooltip" data-placement="top" title="Normalized Effect Size (NES) is defined as the effect of one allele on the inverse-normalized expression level of the associated gene. Follow this link for more details from the GTEx Portal."> Definition of NES <span class="fa fa-external-link"></span> </a>
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" title="Created by Alan Kwong, Mukai Wang, Andy Boughton, Peter VandeHaar, and Hyun Min Kang">
            <span class="badge badge-pill badge-secondary" style="pointer-events: none;"><span class="fa fa-lightbulb-o"></span> Credits </span>
          </span>

        </div>
      </div>

</body>
</html>
